src/internal/Subscription.ts

```diff
@@ -126,51 +110,17 @@ export class Subscription implements SubscriptionLike {
         // execute whatever finalizer is handed to it automatically.
         execFinalizer(teardown);
       } else {
-        if (teardown instanceof Subscription) {
-          // We don't add closed subscriptions, and we don't add the same subscription
-          // twice. Subscription unsubscribe is idempotent.
-          if (teardown.closed || teardown._hasParent(this)) {
-            return;
-          }
-          teardown._addParent(this);
+        if (teardown && 'add' in teardown) {
+          // If teardown is a subscription, we can make sure that if it
+          // unsubscribes first, it removes itself from this subscription.
+          teardown.add(() => {
```

Comment:
No longer keeping track of "parents" internally, instead we're optimizing for a world where child subscriptions remove themselves if they unsubscribe. The side effect to this is if you add and remove a child subscription from a parent subscription a number of times (an extreme edge case), or if you remove a child subscription from a parent, then leak the child subscription for some reason, it will keep a reference to the parent until you unsubscribe it or garbage collect it. These are corner cases, and we're able to remove a lot of weight if we just instruct people to clean up after themselves. :)